src/game/inject.c

```diff
@@ -0,0 +1,721 @@
+        inj_info.anim_bone_count, fp);
+    File_Read(
+        g_AnimFrames + level_info->anim_frame_count, sizeof(int16_t),
+        inj_info.anim_frame_count, fp);
+
+    for (int i = 0; i < inj_info.anim_count; i++) {
+        ANIM_STRUCT *anim = &g_Anims[level_info->anim_count + i];
+
+        File_Read(&anim->frame_ofs, sizeof(uint32_t), 1, fp);
+        File_Read(&anim->interpolation, sizeof(int16_t), 1, fp);
+        File_Read(&anim->current_anim_state, sizeof(int16_t), 1, fp);
+        File_Read(&anim->velocity, sizeof(int32_t), 1, fp);
+        File_Read(&anim->acceleration, sizeof(int32_t), 1, fp);
+        File_Read(&anim->frame_base, sizeof(int16_t), 1, fp);
+        File_Read(&anim->frame_end, sizeof(int16_t), 1, fp);
+        File_Read(&anim->jump_anim_num, sizeof(int16_t), 1, fp);
+        File_Read(&anim->jump_frame_num, sizeof(int16_t), 1, fp);
+        File_Read(&anim->number_changes, sizeof(int16_t), 1, fp);
+        File_Read(&anim->change_index, sizeof(int16_t), 1, fp);
+        File_Read(&anim->number_commands, sizeof(int16_t), 1, fp);
+        File_Read(&anim->command_index, sizeof(int16_t), 1, fp);
+
+        // Re-align to the level.
+        anim->jump_anim_num += level_info->anim_count;
+        anim->frame_ofs += level_info->anim_frame_count * 2;
+        anim->frame_ptr = &g_AnimFrames[anim->frame_ofs / 2];
+        if (anim->number_changes) {
+            anim->change_index += level_info->anim_change_count;
+        }
+        if (anim->number_commands) {
+            anim->command_index += level_info->anim_command_count;
+        }
+    }
+
+    // Re-align to the level.
+    for (int i = 0; i < inj_info.anim_change_count; i++) {
+        g_AnimChanges[level_info->anim_change_count++].range_index +=
+            level_info->anim_range_count;
+    }
+
+    for (int i = 0; i < inj_info.anim_range_count; i++) {
+        g_AnimRanges[level_info->anim_range_count++].link_anim_num +=
+            level_info->anim_count;
+    }
+}
+
+static void Inject_ObjectData(
+    INJECTION *injection, LEVEL_INFO *level_info, RGB888 *palette)
+{
+    INJECTION_INFO inj_info = injection->info;
+    MYFILE *fp = injection->fp;
+
+    for (int i = 0; i < inj_info.object_count; i++) {
+        int32_t tmp;
+        File_Read(&tmp, sizeof(int32_t), 1, fp);
+        OBJECT_INFO *object = &g_Objects[tmp];
+
+        File_Read(&object->nmeshes, sizeof(int16_t), 1, fp);
+        File_Read(&object->mesh_index, sizeof(int16_t), 1, fp);
+        object->mesh_index += level_info->mesh_ptr_count;
+        File_Read(&object->bone_index, sizeof(int32_t), 1, fp);
+        object->bone_index += level_info->anim_bone_count;
+
+        File_Read(&tmp, sizeof(int32_t), 1, fp);
+        object->frame_base =
+            &g_AnimFrames[(tmp + level_info->anim_frame_count * 2) / 2];
+        File_Read(&object->anim_index, sizeof(int16_t), 1, fp);
+        object->anim_index += level_info->anim_count;
+        object->loaded = 1;
```

Comment:
A lot of this code seems similar to the level loader in level.c with occassional follow-up calls to `Inject_` static functions.
Could we maybe share these functions between these two modules, favoring `level.c` and making the `inject` more lightweight? Would it make the code prettier, or is my suggestion ill-formed?

Reply:
I agree, there is a lot of the same kind of code, but my gut feeling was to keep it as separated as possible, mainly because for example on line 408 above, we need to be careful that we're using the right frame index base when injecting. I feel that it might end up needing a lot of additional arguments passed around. Also, thinking ahead, when we come to properly injecting such things as animations into already existing objects (as opposed to injecting the full object set) we'll need to be careful to keep frames, meshes etc contiguous with respect to the objects, so again that kind of thing would need additional information passed around if it were handled in a shared function.
It might be good to come back to this once we have a fuller set of injection types supported maybe? Then we'll know exactly what can be shared out. But if you feel that's not optimal for now, I can work through it and see how the shared function approach will look.

Reply:
Hmm, I agree that this approach leaves us with more flexibility especially that we know more things are planned. Let's try to remember to revisit this later once we have more data. Perhaps a comment `// TODO: consider refactoring once we have more injection use cases`?

Reply:
Ok, that sounds like a good plan. I've added this comment for the time being.